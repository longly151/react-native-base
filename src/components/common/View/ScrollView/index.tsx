import {
  ScrollView as RNScrollView,
  ScrollViewProps as RNScrollViewProps,
  StyleProp,
  ViewStyle,
} from 'react-native';
import React from 'react';
import withTailwind from '@components/hoc/withTailwind';

export interface ScrollViewProps extends RNScrollViewProps {
  className?: string;
}

const ScrollView: React.FC<ScrollViewProps> = React.forwardRef((props, ref: any) => {
  const { style: styleProp, ...otherProps } = props;
  const style: StyleProp<ViewStyle> = [styleProp];
  return (
    <RNScrollView testID="ScrollView" ref={ref} {...otherProps} style={style}>
      {props.children}
    </RNScrollView>
  );
});

export interface ScrollViewType extends RNScrollView {}

export default withTailwind(ScrollView) as React.FC<
  ScrollViewProps & { ref?: React.LegacyRef<RNScrollView> }
>;
