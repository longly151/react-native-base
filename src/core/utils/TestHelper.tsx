import React, { useLayoutEffect } from 'react';
import { render, RenderAPI } from '@testing-library/react-native';
import { Provider } from 'react-redux';
import { PersistGate } from 'redux-persist/integration/react';
import store, { persistor, createStore } from '@configs/store';
import { LoadingScreen } from '@app';
import { NavigationContainer } from '@react-navigation/native';
import Navigation from './Navigation';
import MockAdapter from 'axios-mock-adapter';
import { Axios } from '@utils/Api';

import DefaultColor from '@themes/DefaultColor';
import { RootState, useAppDispatch, useAppSelector } from './Redux';
import _ from 'lodash';

import { renderHook } from '@testing-library/react-hooks';

type DeepPartial<T> = T extends object
  ? {
      [P in keyof T]?: DeepPartial<T[P]>;
    }
  : T;

/**
 * Login Handler
 */
export const successCredential = { email: 'admin@gmail.com', password: '123456' };
export const successLoginResponse = {
  data: {
    token:
      'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MSwic3RhdHVzIjoiQUNUSVZFIiwicm9sZSI6IkFETUlOIiwicm9sZUlkIjoxLCJwZXJtaXNzaW9ucyI6WyJBTEwiXSwidGFibGUiOiJ1c2VycyIsImlkRm9yZWlnbktleSI6InVzZXJJZCIsImlhdCI6MTY1NTQ1NzU5OCwiZXhwIjoxNjYwNjQxNTk4fQ.ZA1dsxArHAjb60HzQgKbkinl49cf5OiDWASiE_Wr0SA',
    id: 1,
    email: 'admin@gmail.com',
    fullName: 'Admin',
    phone: '0327571918',
    avatar:
      'https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcRlUbAyS_643dq_B69jZAlPNW6_Xc7SLELY6SpRsc5OI2wHiiYG&usqp=CAU',
    role: {
      createdAt: '2021-04-19T05:58:23.094Z',
      updatedAt: '2021-04-19T05:58:23.094Z',
      deletedAt: null,
      id: 1,
      name: 'ADMIN',
      description: 'ADMIN',
      permissions: [
        {
          id: 1,
          name: 'ALL',
          description: null,
        },
      ],
    },
  },
};

export const failCredential = { email: 'client@gmail.com', password: '123456' };
export const failLoginResponse = {
  statusCode: 401,
  message: [
    {
      code: 'INCORRECT_EMAIL_PASSWORD',
      description: 'Incorrect email or password',
    },
  ],
  error: 'Unauthorized',
};

const auth = {
  loading: false,
  data: successLoginResponse.data,
  error: null,
};
const withAuthStore = createStore({ auth });

/**
 * Wrapper handler
 */

export type OtherRenderOptions = {
  preloadedState?: DeepPartial<RootState>;
  withAuth?: boolean;
  disableNavigationRef?: boolean;
};

const AppContainer = ({
  children,
  disableNavigationRef,
}: {
  children: any;
  disableNavigationRef?: boolean;
}) => {
  const [loginData] = useAppSelector(state => [state.auth.data]);

  const dispatch = useAppDispatch();

  useLayoutEffect(() => {
    const initApp = () => {
      if (loginData?.token) {
        global.token = loginData.token;
      }
      global.fn = {} as any;
      global.dispatch = dispatch;
    };
    initApp();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  return (
    <NavigationContainer
      ref={disableNavigationRef ? null : Navigation.navigationRef}
      theme={DefaultColor}
    >
      {children}
    </NavigationContainer>
  );
};

const AppProviders = ({
  children,
  preloadedState,
  withAuth,
  disableNavigationRef,
}: {
  children: any;
} & OtherRenderOptions) => {
  const mergedPreloadedState: DeepPartial<RootState> = withAuth
    ? _.merge({ auth }, preloadedState as any)
    : preloadedState;

  return (
    // Create new store if preloadedState !== undefined
    <Provider
      store={preloadedState ? createStore(mergedPreloadedState) : withAuth ? withAuthStore : store}
    >
      <PersistGate loading={LoadingScreen()} persistor={persistor}>
        <AppContainer disableNavigationRef={disableNavigationRef}>{children}</AppContainer>
      </PersistGate>
    </Provider>
  );
};

const customRender = (
  component: React.ReactElement<any>,
  // options?: RenderOptions & OtherRenderOptions,
  options?: OtherRenderOptions,
): RenderAPI => {
  const CustomWrapper = options?.wrapper || React.Fragment;
  const Wrapper = options?.wrapper
    ? ({ children }: any) => (
        <AppProviders
          preloadedState={options?.preloadedState}
          withAuth={options?.withAuth}
          disableNavigationRef={options.disableNavigationRef}
        >
          <CustomWrapper>{children}</CustomWrapper>
        </AppProviders>
      )
    : ({ children }: any) => (
        <AppProviders
          preloadedState={options?.preloadedState}
          withAuth={options?.withAuth}
          disableNavigationRef={options?.disableNavigationRef}
        >
          {children}
        </AppProviders>
      );

  return render(component, { ...options, wrapper: Wrapper });
};

const customRenderHook: typeof renderHook = (callback, options) => {
  const disableNavigationRef =
    typeof options?.disableNavigationRef === 'undefined' ? true : options?.disableNavigationRef;
  const CustomWrapper: any = options?.wrapper || React.Fragment;
  const Wrapper = options?.wrapper
    ? ({ children }: any) => (
        <AppProviders
          preloadedState={options?.preloadedState}
          withAuth={options?.withAuth}
          disableNavigationRef={disableNavigationRef}
        >
          <CustomWrapper>{children}</CustomWrapper>
        </AppProviders>
      )
    : ({ children }: any) => (
        <AppProviders
          preloadedState={options?.preloadedState}
          withAuth={options?.withAuth}
          disableNavigationRef={disableNavigationRef}
        >
          {children}
        </AppProviders>
      );

  return renderHook(callback, { ...options, wrapper: Wrapper });
};

// re-export everything
export * from '@testing-library/react-native';

// override render method
export { customRender as render };

export { customRenderHook as renderHook };

const mockAxios = new MockAdapter(Axios, { delayResponse: 200 });

export { mockAxios };
