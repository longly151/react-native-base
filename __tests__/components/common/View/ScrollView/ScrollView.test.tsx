import React from 'react';
import { render } from '@utils/TestHelper';
import ScrollView from '@components/common/View/ScrollView';

test('should render correctly', () => {
  const { toJSON, getByTestId } = render(<ScrollView />);
  expect(toJSON()).toMatchSnapshot();
  expect(getByTestId('ScrollView')).toBeTruthy();
});

test('should have working props', () => {
  const { getByTestId } = render(
    <ScrollView
      scrollEnabled={false}
      className="flex-1 flex-row self-center justify-between items-center rounded border mx-1 my-1 p-1 border-sky-700 bg-sky-500"
    />,
  );

  expect(getByTestId('ScrollView')).toHaveStyle({
    flexGrow: 1,
    flexShrink: 1,
    flexDirection: 'row',
    alignSelf: 'center',
    justifyContent: 'space-between',
    alignItems: 'center',
    borderBottomLeftRadius: 4,
    borderBottomRightRadius: 4,
    borderTopLeftRadius: 4,
    borderTopRightRadius: 4,
    borderLeftWidth: 1,
    borderRightWidth: 1,
    borderTopWidth: 1,
    borderBottomWidth: 1,
    borderLeftColor: 'rgba(3, 105, 161, 1)',
    borderRightColor: 'rgba(3, 105, 161, 1)',
    borderTopColor: 'rgba(3, 105, 161, 1)',
    borderBottomColor: 'rgba(3, 105, 161, 1)',
    backgroundColor: 'rgba(14, 165, 233, 1)',
    marginLeft: 4,
    marginRight: 4,
    marginTop: 4,
    marginBottom: 4,
    paddingLeft: 4,
    paddingRight: 4,
    paddingTop: 4,
    paddingBottom: 4,
  });

  expect(getByTestId('ScrollView')).toHaveProp('scrollEnabled', false);
});
