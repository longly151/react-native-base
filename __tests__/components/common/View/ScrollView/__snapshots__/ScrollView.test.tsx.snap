// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<RCTScrollView
  style={
    Array [
      Array [
        Object {},
        undefined,
      ],
    ]
  }
  testID="ScrollView"
>
  <View />
</RCTScrollView>
`;
